code_generation/optimizer.py

```diff
@@ -1,77 +1,84 @@
-from code_generation.quantity import NanoAODQuantity
-from code_generation.producer import Filter, BaseFilter
+from __future__ import annotations  # needed for type annotations in > python 3.7
+from code_generation.quantity import NanoAODQuantity, Quantity
+from code_generation.producer import Filter, BaseFilter, Producer, ProducerGroup
+from typing import Set, Tuple, Union, List
 import logging
 
 log = logging.getLogger(__name__)
 
-"""
-Class used to check if the producer ordering is correct,
-If it is not, the Optimize function will auto-correct it.
-Additionally, the optimize attempts to put filters at the top of the list,
-or as far up as possible. A wrong configuration due to missing inputs will also be caught here.
-
-If the scope is not global, the outputs generated by producers in the global scope are also considered.
-"""
-
 
 class ProducerOrdering:
-
     """
-    Init function
+    Class used to check if the producer ordering is correct,
+    If it is not, the Optimize function will auto-correct it.
+    Additionally, the optimize attempts to put filters at the top of the list,
+    or as far up as possible. A wrong configuration due to missing inputs will also be caught here.
 
-    Args:
-        config: The configuration dictionary
-        scope: The scope of the producer ordering
+    If the scope is not global, the outputs generated by producers in the global scope are also considered.
     """
 
-    def __init__(self, config, scope):
-        self.global_producers = config["producers"]["global"]
-        self.ordering = config["producers"][scope]
+    def __init__(
+        self,
+        global_producers: List[Producer | ProducerGroup],
+        scope: str,
+        producer_ordering: List[Producer | ProducerGroup],
```

Comment:
`producer_ordering` is missing in the Args description
https://github.com/KIT-CMS/CROWN/blob/847581d22455fcc82e60e06b8752760b5cfc5e91/code_generation/optimizer.py#L29-L32